File: 287-find-the-duplicate-number/find-the-duplicate-number.java
class Solution {
    public int findDuplicate(int[] nums) {
        int hare = 0;
        int turtle = 0;

        while(true){
            hare = nums[nums[hare]];
            turtle = nums[turtle];
            if (hare == turtle){
                int pointer = 0;
                while(pointer != turtle){
                    pointer = nums[pointer];
                    turtle = nums[turtle];
                }
                return pointer;
            }
        }
        
    }
}

File: 287-find-the-duplicate-number/find-the-duplicate-number.js
var findDuplicate = function(nums) {
    let hare = 0;
    let tortoise = 0;
    while(true){
        hare = nums[nums[hare]];
        tortoise = nums[tortoise];
        if (hare === tortoise){
            let pointer = 0;
            while(pointer !== tortoise){
                pointer = nums[pointer];
                tortoise = nums[tortoise];
            }
            return pointer;
        }
    }
};